packages/element/src/react-platform.native.js

```diff
@@ -8,27 +8,39 @@ import { omit } from 'lodash';
  * WordPress dependencies
  */
 import { applyFilters, doAction } from '@wordpress/hooks';
+import { Component } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import { cloneElement } from './react';
 
-const render = ( element, id ) =>
-	AppRegistry.registerComponent( id, () => ( propsFromParent ) => {
-		const parentProps = omit( propsFromParent || {}, [ 'rootTag' ] );
+const render = ( element, id ) => {
+	class App extends Component {
+		constructor() {
+			super( ...arguments );
 
-		doAction( 'native.pre-render', parentProps );
+			const parentProps = omit( this.props || {}, [ 'rootTag' ] );
 
-		const filteredProps = applyFilters(
-			'native.block_editor_props',
-			parentProps
-		);
+			doAction( 'native.pre-render', parentProps );
```

Comment:
The `native.pre-render` action will be executed right before the component is mounted, this means before the first render.

Reply:
I do not follow exactly how the "ceremony" of a class addresses the original issue of multiple calls to `native.pre-render`. Is `render`/`AppRegistry.registerComponent` still called multiple times? Is the class instance retained between the calls?

Reply:
As far as I checked, the multiple calls are caused when returning a function instead of a component when registering the component in AppRegistry. For example, if you replace the line 41 with the following code:

```
	AppRegistry.registerComponent( id, () => ( props ) => {
		debugger;
		return <App { ...props } />;
	} );
```

The debugger will pause for every warning and error that happens during the initialization, I haven't located yet the exact part from the React Native code that causes this. My impression is that when a warning/error is handled and tries to get the component stack, for some reason is calling the function returned from the callback passed in the `AppRegistry.registerComponent`.

I haven't checked with a functional component, but in this case, since we need to execute the pre-render action on the constructor, I haven't found a reliable way to do it with a functional component so I used a classic component.

Reply:
> The debugger will pause for every warning and error that happens during the initialization, I haven't located yet the exact part from the React Native code that causes this. My impression is that when a warning/error is handled and tries to get the component stack, for some reason is calling the function returned from the callback passed in the `AppRegistry.registerComponent` call.

Ah, I see. From your statement and debugging myself a little further my perception is that the component registered with `AppRegistry` is merely re-rendering when these warnings/errors occur. I presume that is expected, but I have no explicit source to back up that presumption. However, I now understand the `doAction` calls to be side effects and shouldn't be invoked on every render. 

> I haven't checked with a functional component, but in this case, since we need to execute the pre-render action on the constructor, I haven't found a reliable way to do it with a functional component so I used a classic component.

From a brief test, wrapping the side effects with `useEffect` does appear to have the same outcome, i.e. preventing multiple runs, but there is not explicit reason to avoid a class. So, leveraging the class approach is OK with me. 

<details><summary>Functional App</summary>

```jsx
/**
 * External dependencies
 */
import { AppRegistry } from 'react-native';
import { omit } from 'lodash';

/**
 * WordPress dependencies
 */
import { applyFilters, doAction } from '@wordpress/hooks';
import { useEffect } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { cloneElement } from './react';

const render = ( element, id ) =>
	AppRegistry.registerComponent( id, () => ( propsFromParent ) => {
		const parentProps = omit( propsFromParent || {}, [ 'rootTag' ] );

		useEffect( () => {
			doAction( 'native.pre-render', parentProps );
		}, [] );

		const filteredProps = applyFilters(
			'native.block_editor_props',
			parentProps
		);

		useEffect( () => {
			doAction( 'native.render', filteredProps );
		}, [] );

		return cloneElement( element, filteredProps );
	} );

/**
 * Render a given element on Native.
 * This actually returns a componentProvider that can be registered with `AppRegistry.registerComponent`
 *
 * @param {WPElement} element Element to render.
 */
export { render };
```

</details>

In regards to "we need to execute the pre-render action on the constructor," is this to imply that `doAction( 'native.render', filteredProps );` (or other code) is dependent upon `doAction( 'native.pre-render', parentProps );`? Or was your statement merely your interpretation based upon the existing name "pre-render"?

Reply:
> > The debugger will pause for every warning and error that happens during the initialization, I haven't located yet the exact part from the React Native code that causes this. My impression is that when a warning/error is handled and tries to get the component stack, for some reason is calling the function returned from the callback passed in the `AppRegistry.registerComponent` call.
> 
> Ah, I see. From your statement and debugging myself a little further my perception is that the component registered with `AppRegistry` is merely re-rendering when these warnings/errors occur. I presume that is expected, but I have no explicit source to back up that presumption. However, I now understand the `doAction` calls to be side effects and shouldn't be invoked on every render.

Exactly, the `doAction` calls should be treated as side effects and only executed once.

Regarding why the function is being called, it's a bit hard to follow the trace but it's related to how the `react-devtools` package builds the component stacktrace when a warning/error is triggered, here are some screenshots to clarify it:

**1. Log warning is triggered**
<img width="1096" alt="log-warning" src="https://user-images.githubusercontent.com/14905380/125599204-0ab2ee43-b331-46fa-8d36-8f09edbb1cd4.png">

**2. Getting the component stack (`react-dev-tools`)**

The important part here is the `current` variable that holds the current Fiber node and is passed down the stack trace.
<img width="1142" alt="fibernode" src="https://user-images.githubusercontent.com/14905380/125599208-a38f0ee3-4577-4dac-92b4-e4fd958db3b2.png">

Here you can see that the `type` property of the current Fiber node is the App function:
<img width="1132" alt="fibernode-type" src="https://user-images.githubusercontent.com/14905380/125599214-cb68a998-486c-4fdd-a89f-2eb7cdb9c252.png">

**3. Describe function component frame**

The previous `current` value is passed as the `workInProgress` argument in the function shown in the screenshot, note that the first argument of the call to `describeFunctionComponentFrame` is the `type` value (the App function) that was mentioned previosly.
<img width="1120" alt="describe-fiber" src="https://user-images.githubusercontent.com/14905380/125599221-caa6f707-9354-4b2a-9689-4cdeef38c09f.png">

**4. Describe native component frame - App function call**
<img width="1124" alt="fn-call" src="https://user-images.githubusercontent.com/14905380/125599227-1ff3b4dd-bcc0-4a49-add1-85069040fea2.png">
 
> > I haven't checked with a functional component, but in this case, since we need to execute the pre-render action on the constructor, I haven't found a reliable way to do it with a functional component so I used a classic component.
> 
> From a brief test, wrapping the side effects with `useEffect` does appear to have the same outcome, i.e. preventing multiple runs, but there is not explicit reason to avoid a class. So, leveraging the class approach is OK with me.

Thanks for sharing the functional component approach 🙇 ! I agree that wrapping the `doAction` calls with `useEffect` is enough to prevent calling them multiple times, but I decided to go with the class approach because I thought that the `native.pre-render` call should happen before the component is rendered and `useEffect` is triggered after the first render. Due to this reason is why I added the `native.pre-render` action in the class contructor, I found some workarounds to run code on a functional component before the first render but looked like hacks.

> In regards to "we need to execute the pre-render action on the constructor," is this to imply that `doAction( 'native.render', filteredProps );` (or other code) is dependent upon `doAction( 'native.pre-render', parentProps );`? Or was your statement merely your interpretation based upon the existing name "pre-render"?

As far as I checked in [the file where the actions are defined](https://github.com/wordpress-mobile/gutenberg-mobile/blob/864c1a9c2687cb842060aa06be7cab1f957904da/src/index.js#L18-L50), I don't feel there are any dependency between the actions but I infer from the action's name `native.pre-render` that this should be called before the editor is rendered 🤔 .

Reply:
> Regarding why the function is being called, it's a bit hard to follow the trace but it's related to how the `react-devtools` package builds the component stacktrace when a warning/error is triggered [...]

Yeah, I did notice — if I remembering correctly — the original multiple side effect calls issue only occurs if Debug is enabled from the developer menu. 

> As far as I checked in [the file where the actions are defined](https://github.com/wordpress-mobile/gutenberg-mobile/blob/864c1a9c2687cb842060aa06be7cab1f957904da/src/index.js#L18-L50), I don't feel there are any dependency between the actions but I infer from the action's name `native.pre-render` that this should be called before the editor is rendered 🤔 .

Makes sense. I have no data to suggest otherwise at this point. The class constructor approach does ensure it invokes before initial render.

Reply:
> Yeah, I did notice — if I remembering correctly — the original multiple side effect calls issue only occurs if Debug is enabled from the developer menu.

I think it also happens when the Debug is disabled because the reusable blocks are still not showing up in that case. I infer it's being called multiple times because the reusable block is being unregistered when this code gets executed a second time, surprisingly the `capabilities` object is empty and since the reusable block has been already registered, the condition is true. 

https://github.com/WordPress/gutenberg/blob/db76f671b07a0fab33de969b3235ca181c5d3a42/packages/react-native-editor/src/index.js#L60-L66

> Makes sense. I have no data to suggest otherwise at this point. The class constructor approach does ensure it invokes before initial render.

👍